### PR TITLE
Implement mypy fixes across all python files

### PR DESCRIPTION
**Pull request by Repiteo**
Part of figuring out what is causing inconsistencies with my prior PR was needing a more concrete and unambiguous foundation to work with. To that end, this PR adds a *ton* of adjustments to the various tool scripts. Most prominently: type hints are now available on **every single file**. While this did occasionally get a bit wordy (shoutouts to `NinjaPaths` in `ninja_syntax.py`), this ultimately made the expected input and output a lot more parsable

Several functions were adjusted in this process, albeit without changing the fundamental intent. There is, however, one primary exception to this: passing paths can now be done directly! The conversion process will now be handled in `ninja_syntax.py` natively, so the streamlined `Path` syntax can be freely used in `project.py` without worry. What's more, because of the `Union` functionality, strings and paths can be passed interchangably, either in isolation or as a collection, and the ninja builder just handles it

### CODE REVIEW

File: tools/project.py [hunk 1]
```diff
@@ -1,3 +1,5 @@
+#!/usr/bin/env python3
```
Comment (encounter) [inline tools/project.py:1]: Non-executable scripts shouldn't have a shebang
(Same for ninja_syntax.py)

**Review by encounter — COMMENTED** [review]
Apologies for not getting to this sooner! I like these changes. Originally I was skeptical about modifying `ninja_syntax.py`, given that it's vendored from https://github.com/ninja-build/ninja/blob/master/misc/ninja_syntax.py, but I think the readability improvements are worth it, and it doesn't appear to get updates upstream very often 🙂 

My only concern, then, is limiting the supported Python versions by using `typing`. Do you happen to know the minimum supported Python version? 3.7?

**Comment by ribbanya** [comment]
It looks like basic support was introduced in 3.5; not sure about the minimum version of these changes.

If you like, I could externalize the type hints to [stub files](https://mypy.readthedocs.io/en/stable/stubs.html), similar to how typescript adds types to legacy javascript libraries, which would address both the minimum version and the presence of changes to `ninja_syntax.py`.

**Comment by Repiteo** [comment]
Having given this more thought, it *should* be possible to leave `ninja_syntax.py` unchanged (minus hinting) & have the conversions occur before getting passed down. I'll test that locally and see if we can retain the benefits of added readability

**Comment by Repiteo** [comment]
Hmm, well it *works*, but there's quite a lot of `Path` used in `project.py`, so there's two main options I would consider:

1. Bite the bullet and just let `ninja_syntax.py` be custom-tailored to this repo
2. Take inspiration from the ninja repo & swap out `Path` for `os.path` equivalents; this would keep everything string-based

**Comment by encounter** [comment]
Looks great as-is! Thanks for the PR.

Will have to update requirements to say Python 3.8+. Hopefully not too many people are running older versions...

**Comment by ribbanya** [comment]
If it helps, 3.7 [reached EOL](https://devguide.python.org/versions/#python-release-cycle) last year.
